openedx/core/djangoapps/xblock/olx_bundle/__init__.py

```diff
@@ -0,0 +1,346 @@
+                    bundle_uuid=self.bundle_uuid,
+                    block_type=block_type,
+                    olx_path=olx_path,
+                )
+        # If that didn't find the file, the usage ID is not a "top-level" block
+        # in this content library.
+        # Instead, one of the OLX files in this bundle contains something like:
+        # <xblock-include usage="custom-name" definition="html/acdef1234"/>
+        child_usages_map = self.get_usages_map()
+        return child_usages_map.get(usage_id, None)
+
+    def get_usages_map(self):
+        """
+        Get the map of usage_id strings -> BundleDefinitionLocator
+        """
+        _tld, usage_aliases_map = self._scan_bundle_olx()
+        return usage_aliases_map
+
+    def get_top_level_definitions(self):
+        """
+        Get the set of BundleDefinitionLocator items in this bundle that
+        have no parent. (Excludes blocks from linked bundles.)
+        """
+        top_level_definitions, _ua = self._scan_bundle_olx()
+        return top_level_definitions
+
+    def _scan_bundle_olx(self):
+        """
+        Helper method to scan bundles that follow certain conventions and return
+        both a list of "top level" blocks and a map of usage IDs to definitions.
+
+        ** In the future this method can/should be optimized away by performing
+           this computation on commit and writing the result into an olx.map
+           file within the bundle itself **
+
+        ** This method follows the semantics of course and library bundles, but
+           other types of bundles may not follow the same patterns. **
+
+        A course or library bundle consists of OLX files that MUST be named
+            {block_type}/{definition_id}/definition.xml
+        For example:
+            html/intro/definition.xml
+            unit/unit1/definition.xml
+        And it may also have linked bundles which contain additional OLX like:
+            linked_bundle/html/intro/definition.xml
+            linked_bundle/second_link/html/intro/definition.xml
+
+        Any OLX file that represents an XBlock type that can have children may
+        contain <xblock-include> nodes, like this:
+            <unit>
+                <xblock-include definition="html/intro" />
+                <xblock-include definition="html/intro" usage="alernate" />
+                <xblock-include definition="linked_bundle/html/intro" usage="linked_intro" />
+            </unit>
+
+        If the included child is from a different bundle, a usage key *must* be
+        specified (this greatly simplifies the implementation and reduces
+        potential namespace collisions).
+
+        If a block is included via an <xblock-include> directive with a
+        usage="..." alias, and the included block has children, the usage IDs of
+        the child blocks (and their children, and so on) will be prefixed with
+        the usage alias. (Otherwise, if that block was used in multiple places
```

Comment:
1. Will this impact the length of the usage key?
2. Does this mean that moving an XBlock from one parent to another (e.g. from one unit to another) will change its usage key?

Reply:
> Will this impact the length of the usage key?

Yes, the child usage keys will be longer as a result.

> Does this mean that moving an XBlock from one parent to another (e.g. from one unit to another) will change its usage key?

Usually no. But if one of its ancestors is a child of an `<xblock-include usage="...">` directive, then yes. I don't like that but couldn't think of any way around it.